Feature/wsc attrlist

tomereli opened the pull request:
The design done as part of #597 and implemented in #584 provided a generic infrastructure for creating and parsing lists of classes. The reason for this refactor was to provide the infrastructure for supporting a list of wsc attributes in TLVF since currently, prplMesh has YAML files for WSC M1 and WSC M2 TLVs with fixed attributes, which means that if a WSC message received from a *different* controller or agent, with slightly different attributes order and/or less/more attributes, we will fail in parsing.

This PR adds `class AttrList` class which derives from `ClassList` and is used to create and parse WSC attributes. The `AttrList` is a pure virtual class that provides `create()` and `parse()` factory methods which are used to create or parse a WSC attributes list using derived classes - m1 and m2.
The design can be extended to support any WSC message, but for EasyMesh we only need m1 and m2.

The implementation of m1 and m2 is simple and aims to provide full backward compatability to old prplMesh versions since it follows the same order of attributes defined in the tlvWscM2.yaml and tlvWscM1.yaml files.

With this, prplMesh *should* be able to parse any WSC message received from any controller and agent.

Fixes #471 
Fixes #160

Commit:
tlvf: wsc: add m1 & m2 AttrList implementation

AttrList is a pure virtual class derived from ClassList which exposes
factory methods to create or parse an M1 or M2 Attribute list.

On create, we are using a configuration struct in order to not expose
the user to the necessity to add attributes in order.
Attributes are added in the same order as tlvWscM1.yaml and
tlvWscM2.yaml __only__ to be backward compatible with tlvWscM1
parsing in the controller. Once all components align to the new M1 and
M2 classes, the order won't matter but we keep the same order so
prplMesh will be backward compatible to old prplmesh versions which
expect this order.

This is also the reason that create() is not pure virtual -
common attributes to both M1 and M2 which could have been added to a
common function then call the derived class to fill the missing
attributes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: wsc: fix typo in serial number attribute

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: wsc: add missing attributes

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: wsc: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: wsc: use unknown length lists for strings

Update cWscAttrManufacturer, cWscAttrModelNumber, cWscAttrSerialNumber
and cWscAttrModelName to use unknown length lists (list []) for the
string data. This is needed since variable length lists tlvf support
only works for non TLV classes, otherwise the length is incremented
twice. This is probably a bug in TLVF, but it anyway makes more sense to
use unknown length lists here.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: parse M1 using WSC::m1

Switch to using WSC::m1 which automatically parses all attributes in the
M1 using the ClassList approach.
With this commit, the controller can now "talk" with other agents which
creates the M1 message in any attribute order.
This is backward compatible with older prplMesh agents which use the old
tlvWscM1.yaml to create M1 messages.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: remove debug prints which clatter logs

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: create M2 using WSC::m2

Now that we have support for creating m2 as an attribute list which can
be in any order, switch to creating m2 using WSC::m2 instead of
ieee1905_1::tlvWscM2. At this point, since WSC::m2 is creating the exact
same attributes in the exact same order as does tlvWscM2, the agent
which still uses the old parsing using tlvWscM2 still works, which
proves this commit is backward compatible with older prplMesh versions.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: remove use of old WSC TLVs

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: parse M2 using WSC::m2

Switch to using WSC::m2 which automatically parses all attributes in the
M2 message using ClassList approach, so order does not matter.
With this commit, the agent can "talk" with any controller since it no
longer requires the order of attributes to be fixed.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: remove use of old WSC TLVs

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: CmduTlvParser use WSC::AttrList

Remove usage of obsolete tlvWscM1.yaml and tlvWscM2.yaml and use
WSC::AttrList::parse() instead.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: align tlvf_test to WSC::AttrList changes

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: completely remove tlvWscM1.yaml and tlvWscM2.yaml

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [TLVF WSC TLV parser](https://trello.com/c/Hhp90qPO/21-tlvf-wsc-tlv-parser)

Commit:
tlvf: wsc: remove unused attributes

Remove unused attributes from WSC_Attributes.yaml, which have been
replaced with TLVF classes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogen files for  WSC_Attributes.yml

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml:
```diff
@@ -680,3 +680,107 @@ cWscAttrOsVersion:
+    _value: ATTR_MAC_ADDR
+  length:
+    _type: uint16_t
+    _value_const: WSC_MAC_LENGTH
```
Why do we need const values inside the message?
It's part of the EasyMesh specification?

tomereli replied:
A WSC attribute has a length and data, as a TLV does.
Specifying the length as `_value_const` is a tlvf feature which basically sets the `length()` getter to return a const so it can't be changed accidentally.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+ * @brief wsc configuration
+ * 
+ */
+struct config {
```
`config` is a very generic name... I would consider changing this to a more meaningful name (especially since it's in a header file)

tomereli replied:
It is part of the WSC namespace - so the usage is `WSC::config`, don't you think its enough? it is precisely that - a WSC configuration for creating WSC messages. Do you have a better name to suggest?

vitalybu replied:
Depends on the amount of classes in the WSC namespace. If you believe there won't be any other "configuration" structure and/or related data-types it's probably fine.

tomereli replied:
I think its fine, if there will be an issue we can always change it in the future.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+    uint16_t type;
+    uint16_t length;
+    uint8_t *data() { return ((uint8_t *)this + sizeof(*this)); }
+} __attribute__((packed)) sWscAttrHeader;
```
I'm not sure what's the behavior of all compilers when packing a struct that also contains methods...

tomereli replied:
It is not compiler related AFAIK - the only difference between a struct and class in C++ is the default accessibility of member variables and methods. In a struct they are public; in a class they are private, so there shouldn't be any dependency on one compiler or the other.

vitalybu replied:
Right, but you don't use classes the same way you use structs.
For example, you wouldn't `memcpy` into a class or "pack" its' members...
What I meant is that I don't know how the compiler will "pack" the members, since it also contains a method.

tomereli replied:
I see, but I prefer not to change this based on a hunch. We can count on the CI to catch it if there is an issue with it.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+
+public:
+    virtual ~AttrList()                  = default;
+    virtual eWscMessageType type()       = 0;
```
Why is this a pure-virtual method and not say a simple getter method?
Won't it be easier to pass the Type in the constructor and have the base class simply return it?
Anyhow, getter method should probably be defined as const (e.g. `virtual eWscMessageType type() const = 0;`)

tomereli replied:
I guess it could be a simple getter, just like the other common attributes getters defined in the AttrList class.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+    virtual bool valid_custom() = 0;
+
+private:
+    static eWscMessageType wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len);
```
Any reason to not convert this method to static function in the cpp file?
I prefer using stand-alone static functions as it cleans up the header file (which is the actual interface to the class).

vitalybu replied:
Also, the input pointer type should probably be `void*`. Can prevent casting issues.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+    static eWscMessageType wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len);
+};
+
+class m1 : public AttrList {
```
Why did you decide to keep it in the same header file?

tomereli replied:
No specific reason, just for convenience. I'll split to separate headers and files now.

vitalybu commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+    uint8_t *registrar_nonce() { return getAttr<cWscAttrRegistrarNonce>()->nonce(); };
+    cWscAttrEncryptedSettings &encrypted_settings()
+    {
+        return *getAttr<cWscAttrEncryptedSettings>();
```
What if this attribute doesn't exists? You'll dereference a nullptr?
You should either make sure it exists or return the `shared_ptr`

tomereli replied:
The design of these classes is that the `AttrList` is an abstract class with factory methods to create and parse M1 and M2 messages. The `AttrList::create()` uses `wsc_msg_type_from_buffer()` to get the message type and then creates a `shared_ptr` of the appropriate derived class - M1 or M2.
Then it calls one of the 2 versions of the `init()` virtual function which calls `addAttr<T>` for all attributes which belong in this message.
Finally, the `valid()` function validates that the message includes all the mandatory attributes, for those we have the getter functions.
Bottom line - M2 and M1 should __only__ be created using one of the factory methods - 
 `AttrList::create()` or `AttrList::parse()`, which ensures all the attributes are there and can be dereferenced. This makes the rest of the code much simpler, especially for the callers which remain almost identical to how they were before this change.

vitalybu replied:
ACK

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+        if (!attributes->init(cfg))
+            return nullptr;
+        attributes->finalize();
+        tlv.addInnerClassList(attributes);
```
This part is duplicated between the M1 and M2. You can move it outside the if statement.

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+    while (ntohs(attr->type) != ATTR_MSG_TYPE) {
+        attributes_length += ntohs(attr->length) + sizeof(sWscAttrHeader);
+        if (attributes_length > buff_len)
+            return eWscMessageType::WSC_MSG_TYPE_INVALID;
```
Is this the correct error code for this?

tomereli replied:
Yes, why would you think otherwise?

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+    return static_cast<eWscMessageType>(*attr->data());
+}
+
+std::shared_ptr<AttrList> AttrList::parse(ieee1905_1::tlvWsc &tlv)
```
`tlv` isn't modified in this method. Can be changed to `const`.

tomereli replied:
It is modified - we are adding the WSC attributes in the TLV payload, can't be const.

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+    } else if (msg_type == eWscMessageType::WSC_MSG_TYPE_M2) {
+        attributes = std::make_shared<m2>(tlv.payload(), tlv.payload_length(), true);
+        if (!attributes->init())
+            return nullptr;
```
Move outside the if statement

vitalybu replied:
Also, you should make sure that the allocation was successful:
`if (!attributes || !attributes->init())`

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+bool AttrList::valid()
+{
+    if (!getAttr<cWscAttrVersion>())
+        return false;
```
Why there's no logging for errors?

tomereli replied:
Good point, I'll add logging in the valid() methods.

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+        case ATTR_NETWORK_KEY:
+        case ATTR_KEY_WRAP_AUTH:
+        case ATTR_ENCR_TYPE:
+        case ATTR_AUTH_TYPE:
```
Should add our `FALLTHROUGH` macro. Otherwise it can generate build errors on some compilers.

vitalybu commented:
Reviewed up to the "agent: create M1 using WSC::m1" commit

Commit:
common: fix gcc attribute syntax

Fix wrong usage of __attribute((fallthrough)) which should be
__attribute__((fallthrough)) according to
https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add tlvfdefines.h for FALLTHROUGH define

Add FALLTHROUGH support to tlvf.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml:
```diff
@@ -558,7 +558,7 @@ cWscAttrModelName:
   model:
     _type: char
-    _length: [ length ]
+    _length: []
```
This change is fishy, I don't understand how `length` will be updated when allocating the string.

tomereli replied:
As explained oraly, since WSC is also a TLV, it will be updated automatically as its a feature of unknown length lists in TLVF TLV classes.

morantr commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+    std::string manufacturer() { return getAttr<cWscAttrManufacturer>()->manufacturer_str(); };
+    std::string model_name() { return getAttr<cWscAttrModelName>()->model_str(); };
+    std::string device_name() { return getAttr<cWscAttrDeviceName>()->device_name_str(); };
+    std::string serial_number() { return getAttr<cWscAttrSerialNumber>()->serial_number_str(); };
```
Can you change the return value to `const std::string &`?

tomereli replied:
No:
```
./framework/tlvf/src/include/tlvf/WSC/AttrList.h: In member function 'const string& WSC::AttrList::serial_number() const':                                                                                 
../framework/tlvf/src/include/tlvf/WSC/AttrList.h:74:106: error: returning reference to temporary [-Werror=return-local-addr]                                                                               
     const std::string& serial_number() const { return getAttr<cWscAttrSerialNumber>()->serial_number_str(); };  
```

This is because the `manufacturer_str()` returns a temporary std::string() object:

```
std::string cWscAttrManufacturer::manufacturer_str() {
    char *manufacturer_ = manufacturer();
    if (!manufacturer_) { return std::string(); }
    return std::string(manufacturer_, m_manufacturer_idx__);
}
```

tomereli replied:
I did however go over the code and added const to functions which can be const

morantr commented on framework/tlvf/src/include/tlvf/WSC/AttrList.h:
```diff
@@ -0,0 +1,139 @@
+        return reinterpret_cast<sWscAttrHeader *>(
+            m_class_vector.empty() ? getMessageBuff() : m_class_vector.back()->getBuffPtr());
+    }
+    uint16_t getNextAttrType() { return ntohs(getNextAttrHdr()->type); };
```
I think it would be nicer if we use `swap_16()` from "swap.h" instead of `ntohs`.
Same on function `AttrList::wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len)`

tomereli replied:
Actually I really hate that we use swap_* instead of normal linux API ntohs and friends. Using the `ntohX` version makes it perfectly clear that we are changing endianess to/from network/host byte order and not just swapping bytes without a reason. This is why these APIs exist..

morantr commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+        if (!attributes->init(cfg))
+            return nullptr;
+        attributes->finalize();
+        tlv.addInnerClassList(attributes);
```
Need to add a check that the memory has been successfully allocated:
`if (!attributes || !attributes->init(cfg)) { return nullptr; }`
This should be done in several places on this file. Do it in all the places you use `make_shared`.

tomereli replied:
Done

morantr commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+    auto enrolee_nonce_attr = addAttr<WSC::cWscAttrEnrolleeNonce>();
+    if (!enrolee_nonce_attr)
+        return false;
+    std::copy(cfg.enrollee_nonce, cfg.enrollee_nonce + WSC_NONCE_LENGTH,
```
If you will use `std::copy_n(src, len, dest)` it would look nices since you will not have to do `cfg.enrollee_nonce + WSC_NONCE_LENGTH`, instead replace it with just `WSC_NONCE_LENGTH`.
(nitpick)

morantr commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,504 @@
+    if (!getAttr<cWscAttrAuthenticator>())
+        return false;
+    return false;
+}
```
missing last line break

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4569,65 +4571,45 @@ bool slave_thread::add_radio_basic_capabilities()
+    cfg.mac      = network_utils::mac_from_string(backhaul_params.bridge_mac);
+    dh           = std::make_unique<mapf::encryption::diffie_hellman>();
+    std::copy(dh->nonce(), dh->nonce() + dh->nonce_length(), cfg.enrollee_nonce);
+    std::copy(dh->pubkey(), dh->pubkey() + dh->pubkey_length(), cfg.pub_key);
```
std::copy_n (nitpick)

tomereli replied:
Was already like that so I prefer not to change it as part of this PR.

tomereli commented:
> Reviewed up to the "agent: create M1 using WSC::m1" commit

I rebased and squashed with fixups to all your and Moran's comments, leaving open so you can finish reviewing tomorrow.

Commit:
agent: create M1 using WSC::m1

Now that we have support for creating m1 as an attribute list which can
be in any order, first switch to creating m1 using WSC::m1 instead of
ieee1905_1::tlvWscM1. At this point, since WSC::m1 is creating the exact
same attributes in the exact same order as does tlvWscM1, the controller
which still uses the old parsing using tlvWscM1 still works, which
proves this commit is backward compatible with older prplMesh versions.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on framework/tlvf/src/src/WSC/AttrList.cpp:
```diff
@@ -0,0 +1,344 @@
+                return false;
+            }
+            break;
+        // TODO these are not expected attributes (part of config data)
```
Although I commented on the missing FALLTHROUGH attribute, maybe it's better to simply remove unexpected cases.
All unexpected attributes will be caught by the "default" case anyway.

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -843,23 +840,21 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(const std::string &sr
             break;
         }
-        if (!autoconfig_wsc_add_m2(tlvwscM1, &bss_info_conf)) {
+        if (!autoconfig_wsc_add_m2(*m1, &bss_info_conf)) {
```
Why not change the argument of this method to reference as well?

tomereli replied:
m1 is created in this method, it is not passed as an argument.

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3622,23 +3622,23 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
+bool slave_thread::autoconfig_wsc_calculate_keys(WSC::m2 &m2, uint8_t authkey[32],
+                                                 uint8_t keywrapkey[16])
 {
     auto mac = network_utils::mac_from_string(backhaul_params.bridge_mac);
```
No need to "waste" time on converting the MAC address if the DH member is not initialized

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3854,7 +3853,7 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
 
     std::shared_ptr<wfa_map::tlvApRadioIdentifier> ruid = nullptr;
-    std::vector<std::shared_ptr<ieee1905_1::tlvWscM2>> m2_list;
+    std::vector<std::shared_ptr<WSC::m2>> m2_list;
```
As a general rule of thumb, unless there's a need for index based access to elements, it's better to use `std::list` as it's more predictable in runtime complexity and memory allocations.

tomereli replied:
Will change it in a separate commit.

vitalybu commented on framework/tlvf/test/tlvf_test.cpp:
```diff
@@ -235,46 +236,58 @@ bool add_encrypted_settings(std::shared_ptr<tlvWscM2> m2, uint8_t *keywrapkey, u
-    if (!mapf::encryption::aes_encrypt(keywrapkey, iv, encrypted, datalen, encrypted, cipherlen)) {
+    if (!mapf::encryption::aes_encrypt(keywrapkey, cfg.iv, encrypted, datalen, encrypted,
+                                       cipherlen)) {
         LOG(DEBUG) << "aes encrypt";
```
That's an interesting ERROR message :)

Commit:
agent: wsc autoconfig use list instead of vector

https://github.com/prplfoundation/prplMesh/pull/602#discussion_r361399045

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
